platforms/software/dependency-management/src/main/java/org/gradle/api/internal/artifacts/transform/DefaultTransformInvocationFactory.java

```diff
@@ -141,9 +146,15 @@ public Deferrable<Try<ImmutableList<File>>> createInvocation(
                 );
             }
         }
-        return effectiveEngine.createRequest(execution)
-            .executeDeferred(identityCache)
-            .map(result -> result
+        Deferrable<CacheResult<TransformWorkspaceResult>> deferredResult = effectiveEngine.createRequest(execution)
+            .executeDeferred(identityCache);
+        deferredResult.getCompleted()
+            .ifPresent(cacheResult -> cacheResult.getOriginMetadata()
+                .ifPresent(originMetadata -> progressEventEmitter.emitNowIfCurrent(
+                    new DefaultSkippedTransformExecutionProgressDetails(cacheResult, originMetadata))
+                ));
+        return deferredResult
+            .map(result -> result.getResult()
```

Comment:
❓ How about this?

```suggestion
        return effectiveEngine.createRequest(execution)
            .executeDeferred(identityCache)
            .map(result -> {
                result.getOriginMetadata()
                    .ifPresent(originMetadata -> progressEventEmitter.emitNowIfCurrent(
                        new DefaultSkippedTransformExecutionProgressDetails(result, originMetadata))
                    );
                return result;
            })
            .map(result -> result.getResult()
                .map(successfulResult -> successfulResult.resolveForInputArtifact(inputArtifact))
                .mapFailure(failure -> new TransformException(String.format("Execution failed for %s.", execution.getDisplayName()), failure)));
```

I don't think we should rely on `getCompleted()` here, or assume that a skipped result will arrive immediately, no?

I guess what we would need here is something similar to `Stream.peek()`.

Reply:
That is the point here: When we actually need to run something, I don't want the event. Only when it already ran in the previous build. It is somewhat hard to decide that on the Gradle side without capturing some state, we will make that decision in the Develocity plugin.

Note that checking if the origin build id is the same as the current build id does not work, since when the execution actually re-uses something from the build cache or is up-to-date, the origin build will be different although it was executed in the current build.

Reply:
One solution for that could be to store the build invocation id when something has entered the cache in the `IdentityCacheResult`. Then I could compare that build invocation id here with the build invocation id of the current build and only emit the progress event when they are different.